Aplib.Core/Action.cs

```diff
@@ -0,0 +1,36 @@
+﻿using System;
+
+namespace Aplib.Core
+{
+    /// <summary>
+    /// Describes an action that can be executed and guarded.
+    /// </summary>
+    /// <typeparam name="TQuery">The type of the query of the action</typeparam>
+    public class Action<TQuery>
+    {
+        private TQuery? _storedResult;
+
+        /// <summary>
+        /// The effect of the action.
+        /// </summary>
+        public System.Action<TQuery>? Effect { private get; set; }
+
+        public bool Actionable => Query!.Invoke() is not (false or null);
+
+        /// <summary>
+        /// The query of the action. Can return a value, which is stored and can be used in the effect.
+        /// </summary>
+        /// <remarks>A boolean value can also be used. If an object is used, use null to mark the action as unqualified.</remarks>
+        public Func<TQuery>? Query { private get; set; }
+
+        /// <summary>
+        /// Execute the action against the world.
+        /// </summary>
+        public void Execute() => Effect?.Invoke(_storedResult!);
```

Comment:
`_storedResult` might be `null`, so I think it doesn't really make sense to provide this to the `Effect` delegate on invoke, when really it should not execute the `Effect` at all when the result has not been queried.

Edge case to consider: when you make `TQuery` a nullable type, and assign null to `_storedResult`, that would then not execute the `Effect`, while you might actually mean that the _value_ of `_storedResult` is `null`, if that makes any sense. In other words, `_storedResult` being `null` should be different than the result of the `Query` being `null`.

Reply:
I understand that `_storedResult` is not meant to be null and will be overridden when queried. However, right now, the query is not implicit, so one might call `Execute` without calling `Guard` and `_storedResult` would be `null`. This would be treated in the same way as the query being called and returning `null`. The first one **shouldn't** happen, the second one **should** be possible.